### PR TITLE
Feature clean

### PR DESCRIPTION
**Pull request by mnkg561**
close # https://github.com/keikoproj/iam-manager/issues/16

**Could you share the solution in high level?**
This PR includes following items
1. Reconcile logic is re-written to make sure updating status doesn't result in infinite loop events.
2. For READY state, logic is included to verify the status with state of the world(external AWS IAM Policy) to make sure if there is any change to proceed further.
3. Cleaned up troubleshooting log statements.

**Could you share the test results?**
1. Controller restart doesn't result in reconcile again.
2. Updating status to READY at the end of the job doesn't result in external reconcile process.

```
mtvl15367e28a:playerdb nmogulla$ k logs -f -l app=iam-manager
2020-01-29T03:17:55.067Z	INFO	controller-runtime.metrics	metrics server is starting to listen	{"addr": ":8080"}
2020-01-29T03:17:55.068Z	INFO	controller-runtime.controller	Starting EventSource	{"controller": "iamrole", "source": "kind source: /, Kind="}
setting up k8s client
2020-01-29T03:17:55.073Z	INFO	setup	I'm enabling webhook now
2020-01-29T03:17:55.074Z	INFO	controller-runtime.builder	Registering a mutating webhook	{"GVK": "iammanager.keikoproj.io/v1alpha1, Kind=Iamrole", "path": "/mutate-iammanager-keikoproj-io-v1alpha1-iamrole"}
2020-01-29T03:17:55.074Z	INFO	controller-runtime.builder	Registering a validating webhook	{"GVK": "iammanager.keikoproj.io/v1alpha1, Kind=Iamrole", "path": "/validate-iammanager-keikoproj-io-v1alpha1-iamrole"}
2020-01-29T03:17:55.074Z	INFO	setup	starting manager
2020-01-29T03:17:55.074Z	INFO	controller-runtime.manager	starting metrics server	{"path": "/metrics"}
2020-01-29T03:17:55.175Z	INFO	controller-runtime.certwatcher	Updated current TLS certificate
2020-01-29T03:17:55.175Z	INFO	controller-runtime.certwatcher	Starting certificate watcher

2020-01-29T03:18:11.872Z	INFO	controller-runtime.controller	Starting Controller	{"controller": "iamrole"}
#### Controller restart start ####
2020-01-29T03:18:11.872Z	DEBUG	controller-runtime.manager.events	Normal	{"object": {"kind":"ConfigMap","namespace":"iam-manager-system","name":"controller-leader-election-helper","uid":"19d6e103-3726-11ea-bd19-0225db228e40","apiVersion":"v1","resourceVersion":"567703456"}, "reason": "LeaderElection", "message": "iam-manager-controller-manager-667969fdfc-gg5qn_f0f0079a-4245-11ea-a13f-72287ec9f597 became leader"}
2020-01-29T03:18:11.972Z	INFO	controller-runtime.controller	Starting workers	{"controller": "iamrole", "worker count": 1}
2020-01-29T03:18:11.972Z	INFO	controllers.Iamrole	Start of the request	{"request_id": "4a102be2-280b-45a1-b698-9009828b4165", "iamrole": "dev-patterns-iksmanager-usw2-ppd-idev/iamrole-sample2"}
2020-01-29T03:18:11.973Z	INFO	controllers.Iamrole	And the state is 	{"request_id": "4a102be2-280b-45a1-b698-9009828b4165", "iamrole": "iamrole-sample2", "state": "Ready"}
2020-01-29T03:18:12.646Z	INFO	controllers.Iamrole	<B>No change in the incoming policy compare to state of the world(external AWS IAM) policy	{"request_id": "4a102be2-280b-45a1-b698-9009828b4165", "iamrole": "iamrole-sample2"} </B>
2020-01-29T03:18:12.646Z	INFO	controllers.Iamrole	Successfully reconciled	{"request_id": "4a102be2-280b-45a1-b698-9009828b4165", "iamrole": "iamrole-sample2"}
2020-01-29T03:18:12.646Z	DEBUG	controller-runtime.controller	Successfully Reconciled	{"controller": "iamrole", "request": "dev-patterns-iksmanager-usw2-ppd-idev/iamrole-sample2"}
2020-01-29T03:18:12.646Z	INFO	controllers.Iamrole	Start of the request	{"request_id": "d1d076bc-ebdd-4d9b-aaa7-74f9a15933cc", "iamrole": "cert-manager/iamrole-sample2"}
2020-01-29T03:18:12.646Z	INFO	controllers.Iamrole	And the state is 	{"request_id": "d1d076bc-ebdd-4d9b-aaa7-74f9a15933cc", "iamrole": "iamrole-sample2", "state": "Ready"}
2020-01-29T03:18:12.826Z	INFO	controllers.Iamrole	No change in the incoming policy compare to state of the world(external AWS IAM) policy	{"request_id": "d1d076bc-ebdd-4d9b-aaa7-74f9a15933cc", "iamrole": "iamrole-sample2"}
2020-01-29T03:18:12.826Z	INFO	controllers.Iamrole	Successfully reconciled	{"request_id": "d1d076bc-ebdd-4d9b-aaa7-74f9a15933cc", "iamrole": "iamrole-sample2"}
2020-01-29T03:18:12.826Z	DEBUG	controller-runtime.controller	Successfully Reconciled	{"controller": "iamrole", "request": "cert-manager/iamrole-sample2"}
2020-01-29T03:18:48.007Z	INFO	iamrole-resource	default	{"name": "iamrole-sample2"}
2020-01-29T03:18:48.010Z	INFO	iamrole-resource	validate update	{"name": "iamrole-sample2"}

2020-01-29T03:18:48.018Z	INFO	controllers.Iamrole	Start of the request	{"request_id": "8e3f7cd0-d371-414e-beb3-f06d89de76e1", "iamrole": "cert-manager/iamrole-sample2"}
2020-01-29T03:18:48.018Z	INFO	controllers.Iamrole	And the state is 	{"request_id": "8e3f7cd0-d371-414e-beb3-f06d89de76e1", "iamrole": "iamrole-sample2", "state": "Ready"}
2020-01-29T03:18:48.146Z	INFO	controllers.Iamrole	Update request event. Proceeding with update operation	{"request_id": "8e3f7cd0-d371-414e-beb3-f06d89de76e1", "iamrole": "iamrole-sample2"}
EntityAlreadyExists EntityAlreadyExists: Role with name k8s-cert-manager already exists.
	status code: 409, request id: b0ae6dab-2437-46b8-8e1c-72c02e49113f
{

}
2020-01-29T03:18:48.636Z	INFO	controllers.Iamrole	Successfully reconciled	{"request_id": "8e3f7cd0-d371-414e-beb3-f06d89de76e1", "iamrole": "iamrole-sample2"}
2020-01-29T03:18:48.636Z	DEBUG	controller-runtime.controller	Successfully Reconciled	{"controller": "iamrole", "request": "cert-manager/iamrole-sample2"}
2020-01-29T03:18:48.636Z	INFO	controllers.Iamrole	Start of the request	{"request_id": "715a4e2a-cd60-4744-a8c9-fc1675dcad8b", "iamrole": "cert-manager/iamrole-sample2"}
2020-01-29T03:18:48.636Z	INFO	controllers.Iamrole	And the state is 	{"request_id": "715a4e2a-cd60-4744-a8c9-fc1675dcad8b", "iamrole": "iamrole-sample2", "state": "Ready"}
2020-01-29T03:18:48.733Z	INFO	controllers.Iamrole	<B>No change in the incoming policy compare to state of the world(external AWS IAM) policy	{"request_id": "715a4e2a-cd60-4744-a8c9-fc1675dcad8b", "iamrole": "iamrole-sample2"}</B>
2020-01-29T03:18:48.733Z	INFO	controllers.Iamrole	Successfully reconciled	{"request_id": "715a4e2a-cd60-4744-a8c9-fc1675dcad8b", "iamrole": "iamrole-sample2"}
2020-01-29T03:18:48.733Z	DEBUG	controller-runtime.controller	Successfully Reconciled	{"controller": "iamrole", "request": "cert-manager/iamrole-sample2"}
```

### CODE REVIEW

**Comment by mnkg561** [comment]
@kshamajain99 Review this one

File: config/manager/manager.yaml [hunk 1]
```diff
@@ -1,8 +1,6 @@
 apiVersion: v1
 kind: Namespace
 metadata:
-  annotations:
-    iks.intuit.com/managed: "false"
```
Comment (kshamajain99) [inline config/manager/manager.yaml]: Why are we removing this?

Comment (kshamajain99) [inline config/manager/manager.yaml]: Why are we removing this?

File: controllers/iamrole_controller.go [hunk 2]
```diff
@@ -75,16 +79,18 @@ func (r *IamroleReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 		if iamRole.Status.RetryCount != 0 {
 			iamRole.Status.RetryCount = iamRole.Status.RetryCount + 1
 		}
-		log.Info("iam role name in delete", "role", roleName)
+		log.Info("Iamrole delete request")
```
Comment (kshamajain99) [inline controllers/iamrole_controller.go]: I think we should also print rolename in the log here for easier debugging

File: controllers/iamrole_controller.go [hunk 3]
```diff
@@ -107,76 +113,97 @@ var roleTrust = `{
...
+	log := r.Log.WithValues("request_id", ctx.Value(requestId), "iamrole", iamRole.Name)
+	log.Info("And the state is ", "state", iamRole.Status.State)
 	role, _ := json.Marshal(iamRole.Spec.PolicyDocument)
-	fmt.Println(string(role))
 	roleName := fmt.Sprintf("k8s-%s", iamRole.ObjectMeta.Namespace)
 	input := awsapi.IAMRoleRequest{
 		Name:             roleName,
-		PolicyName:       fmt.Sprintf("k8s-%s-policy", iamRole.ObjectMeta.Namespace),
-		Description:      "bla bla",
+		PolicyName:       "custom",
```
Comment (kshamajain99) [inline controllers/iamrole_controller.go]: I think we should create an issue to manage all constants in one place. I might become difficult as the code grows.

File: config/manager/manager.yaml (hunk 1 above)
Comment (mnkg561) [reply config/manager/manager.yaml]: @kshamajain99 Intuit specific annotations. so we shouldn't have that in this repo

Comment (mnkg561) [reply config/manager/manager.yaml]: Same as above

File: controllers/iamrole_controller.go (hunk 3 above)
Comment (mnkg561) [reply controllers/iamrole_controller.go]: Agreed. Lets make sure we handle this in https://github.com/keikoproj/iam-manager/issues/18 along with properties

File: controllers/iamrole_controller.go (hunk 2 above)
Comment (mnkg561) [reply controllers/iamrole_controller.go]: roleName is there and it is coming from the first line of this method. requestId and roleName will be printed in all the log statements in this method

Comment (mnkg561) [reply controllers/iamrole_controller.go]: Deleted here to avoid redundancy.

**Pull request merged**